File: walletcore/src/main/java/com/mycelium/wapi/wallet/eth/EthAddress.java
package com.mycelium.wapi.wallet.eth;

import com.mrd.bitlib.model.AddressType;
import com.mycelium.wapi.wallet.GenericAddress;
import com.mycelium.wapi.wallet.coins.CryptoCurrency;
import com.mycelium.wapi.wallet.eth.coins.EthMain;

import org.jetbrains.annotations.NotNull;

public class EthAddress implements GenericAddress {

    private String address;

    EthAddress(String address) {
        this.address = address;
    }

    @Override
    public CryptoCurrency getCoinType() {
        return EthMain.INSTANCE;
    }

    @Override
    public long getId() {
        return 0;
    }

    public static EthAddress from(CryptoCurrency currencyType, String address) {
        return new EthAddress(address);
    }

    @NotNull
    @Override
    public byte[] getBytes() {
        return new byte[0];
    }
}

File: walletcore/src/main/java/com/mycelium/wapi/wallet/eth/EthAddress.kt
package com.mycelium.wapi.wallet.eth

import com.mrd.bitlib.model.hdpath.HdKeyPath
import com.mycelium.wapi.wallet.Address
import com.mycelium.wapi.wallet.coins.CryptoCurrency
import java.io.IOException
import java.io.ObjectInputStream
import org.web3j.abi.datatypes.Address as E3jAddress


class EthAddress(cryptoCurrency: CryptoCurrency, val addressString: String) : Address {
    @Transient
    var address = E3jAddress(addressString)
    override val coinType = cryptoCurrency

    override fun getSubType() = "default"

    override fun getBytes() = address.toUint().toString().toByteArray()

    override fun toString() = addressString

    /**
     * Always treat de-serialization as a full-blown constructor, by
     * validating the final state of the de-serialized object.
     */
    @Throws(ClassNotFoundException::class, IOException::class)
    private fun readObject(inputStream: ObjectInputStream) {
        //always perform the default de-serialization first
        inputStream.defaultReadObject()

        address = E3jAddress(addressString)
    }

    override fun equals(other: Any?): Boolean {
        if (other !is EthAddress) {
            return false
        }
        return getBytes().contentEquals(other.getBytes())
    }

    override fun hashCode(): Int {
        return getBytes().contentHashCode()
    }

    companion object {
        fun getDummyAddress(cryptoCurrency: CryptoCurrency) =
                EthAddress(cryptoCurrency, "0x000000000000000000000000000000000000dEaD")
    }

    private var _bip32Path: HdKeyPath? = null
    override fun getBip32Path(): HdKeyPath? = _bip32Path

    override fun setBip32Path(bip32Path: HdKeyPath?) {
        _bip32Path = bip32Path
    }
}